astropy/cosmology/core.py

```diff
@@ -198,6 +198,37 @@ def clone(self, *, meta=None, **kwargs):
         # Return new instance, respecting args vs kwargs
         return self.__class__(*ba.args, **ba.kwargs)
 
+    # -----------------------------------------------------
+
+    def __eq__(self, other):
+        """Check equality on all immutable fields.
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.Cosmology` subclass instance
+            The object in which to compare.
+
+        Returns
+        -------
+        bool
+            True if all immutable fields are equal, False otherwise.
+
+        """
+        if not isinstance(other, Cosmology):
+            return False
+
+        sias, oias = self._init_arguments, other._init_arguments
+
+        # check if the cosmologies have identical signatures.
+        # this protects against one cosmology having a superset of input
+        # parameters to another cosmology.
+        if set(sias.keys()).difference(oias.keys()):
```

Comment:
This is not correct: if `oias` has more keys, the result will be empty; just do `if sias.keys() != oias.keys()` (this is not sensitive to order).

Probably good to test this case explicitly!

p.s. If `meta` is ignored below, should it be ignored here too? If so,
`if (sios.kays() ^ oias.keys()) - {'meta'}:` (this `set` symmetric difference operation works on dict keys, it seems, though not sure how long this has been the case)

Reply:
Thanks for the suggestion. I had assumed that since dictionaries are now ordered, so too were the keys. Good to know.

I guess it's possible for a user-defined Cosmology to not have "meta" in the constructor signature, so explicitly excluding it will avoid that rare edge case. I'll check the python version for symmetric difference.